crates/op-revm/tests/test_utils.rs

```diff
@@ -0,0 +1,145 @@
+use std::fs;
+use std::path::PathBuf;
+use serde_json::{json, Value};
+use revm::context_interface::result::ResultAndState;
+use std::io;
+
+/// Compare the EVM execution output with expected output from a testdata file.
+/// If the testdata file doesn't exist, it will be created with the current output.
+/// On mismatch, it will print both expected and actual outputs for easier debugging.
+///
+/// # Arguments
+///
+/// * `testdata_path` - Path to the testdata file relative to the tests directory
+/// * `output` - The output from EVM execution
+///
+/// # Returns
+///
+/// `Ok(())` if the comparison succeeded or the testdata file was created
+/// `Err(io::Error)` if there was an error reading or writing the testdata file
+/// `Err(String)` if there was a mismatch between the expected and actual outputs
+pub(crate) fn compare_or_save_testdata<HaltReasonTy>(
+    testdata_path: &str,
+    output: &ResultAndState<HaltReasonTy>,
+) -> Result<(), Box<dyn std::error::Error>> 
+where
+    HaltReasonTy: serde::Serialize,
+{
+    let tests_dir = PathBuf::from("tests/testdata");
+    let testdata_file = tests_dir.join(testdata_path);
+    
+    // Create directory if it doesn't exist
+    if let Some(parent) = testdata_file.parent() {
+        if !parent.exists() {
+            fs::create_dir_all(parent)?;
+        }
+    }
+    
+    // Serialize the output to JSON
+    let output_json = json!({
+        "result": output.result,
+        "state": output.state
+    });
+    
+    // Check if testdata file exists
+    if !testdata_file.exists() {
+        // Create the testdata file with the current output
+        fs::write(&testdata_file, serde_json::to_string_pretty(&output_json)?)?;
+        println!("Created testdata file: {}", testdata_file.display());
+        return Ok(());
+    }
+    
+    // Read the expected output from the testdata file
+    let expected_json: Value = serde_json::from_str(&fs::read_to_string(&testdata_file)?)?;
```

Comment:
It should be decoded to `ResultAndState` and compared directly with the object. Comparing strings can be flaky as json can contain a hashmap that can have items in arbitrary order.

Reply:
Done